command/agent/http.go

```diff
@@ -191,93 +192,124 @@ func (s *HTTPServer) Shutdown() {
 	}
 }
 
+// handleFuncMetrics takes the given pattern and handler and wraps to produce
+// metrics based on the pattern and request.
+func (s *HTTPServer) handleFuncMetrics(pattern string, handler func(http.ResponseWriter, *http.Request)) {
+	// Get the parts of the pattern. We omit any initial empty for the
+	// leading slash, and put an underscore as a "thing" placeholder if we
+	// see a trailing slash, which means the part after is parsed. This lets
+	// us distinguish from things like /v1/query and /v1/query/<query id>.
+	var parts []string
```

Comment:
```go
const defaultMaxNumParts = 4
parts := make([]string, 0, defaultMaxNumParts)
```

and size `defaultMaxNumParts` to whatever it is that we pass in as the longest URI.  Maybe 8?  But this avoids reallocating and doubling twice for the common case.